### PR TITLE
Refining task api, etc

### PR DESCRIPTION
**Pull request by kywch**
* separated Predicate (`predicate_api.py`) and Task (`task_api.py`)
  * Predicate evaluates progress toward goal and returns float [0,1]
  * Task computes rewards for agents and tracks goal completion
* refactored task api, allow two methods to provide tasks to the env
  * `env.reset(new_tasks` = a list of task instances)
  * `env.reset(teams: Dict[int,List[int]], task_spec` = a list of tasks to instantiate inside the env)
* all the tests are using the new task api

* merged the team helpers into `lib/team_helper.py`
* added event number and gold constraints to predicate

* fixed `IMMORTAL` bug (mainly for performance testing)
* fixed `test_render_save.py`, using the new replay helper

### CODE REVIEW

File: nmmo/core/env.py [hunk 1]
```diff
@@ -131,38 +115,22 @@ def action_space(self, agent):
...
 
     Loads a new game map and returns initial observations
 
     Args:
-        idx: Map index to load. Selects a random map by default
-
+        map_id: Map index to load. Selects a random map by default
+        seed: random seed to use
+        new_tasks: A list of instantiated tasks
+        task_spec: A list of task spec to instantiate inside reset()
```
Comment (daveey) [inline nmmo/core/env.py]: how about just make_new_tasks_fn

that way env doesn't need to know anything about making tasks, or teams

File: nmmo/core/env.py [hunk 2]
```diff
@@ -88,12 +78,6 @@ def box(rows, cols):
     if self.config.PROVIDE_ACTION_TARGETS:
       obs_space['ActionTargets'] = self.action_space(None)
 
-    if self._task_encoding:
```
Comment (daveey) [inline nmmo/core/env.py:91]: don't we need this somewhere?

File: nmmo/core/env.py [hunk 3]
```diff
@@ -186,16 +154,23 @@ def reset(self, map_id=None, seed=None, options=None):
...
+         * task_spec and teams: these are used to instantiate tasks here
+       If these are all None, then use the default task
+    """
+    if new_tasks is not None:
+      # providing an empty new_tasks [] is also possible
+      self.tasks = new_tasks
+    elif task_spec is not None and teams is not None:
+      self.tasks = make_team_tasks(teams, task_spec)
+    else:
+      self.tasks = nmmo_default_task(self.possible_agents)
```
Comment (daveey) [inline nmmo/core/env.py]: this is going to replace existing tasks with the default_task if we don't pass in new_tasks, which is not expected. how about we set self.tasks to default_task in the constructor, and then don't do that here

File: nmmo/core/env.py [hunk 4]
```diff
@@ -308,11 +283,7 @@ def step(self, actions: Dict[int, Dict[str, Dict[str, Any]]]):
 
     # Store the observations, since actions reference them
     self.obs = self._compute_observations()
-    gym_obs = {}
-    for a, o in self.obs.items():
```
Comment (daveey) [inline nmmo/core/env.py:312]: don't we still need this?

File: nmmo/systems/skill.py [hunk 5]
```diff
@@ -265,13 +265,13 @@ def update(self):
...
       return
 
+    if config.IMMORTAL:
+      return
+
     depletion = config.RESOURCE_DEPLETION_RATE
     water = self.entity.resources.water
     water.decrement(depletion)
 
-    if self.config.IMMORTAL:
```
Comment (daveey) [inline nmmo/systems/skill.py:272]: i don't love IMMORTAL, maybe we can just get rid of it?

File: nmmo/task/base_predicates.py [hunk 6]
```diff
@@ -43,6 +43,11 @@ def StayAlive(gs: GameState,
   """True if all subjects are alive.
   """
   return count(subject.health > 0) == len(subject)
+  # The below is for speed testing (bypass GroupView)
```
Comment (daveey) [inline nmmo/task/base_predicates.py]: let's not include it in this PR, perf testing can go in a different PR if we want to commit it at all

File: nmmo/task/base_predicates.py [hunk 7]
```diff
@@ -87,7 +92,7 @@ def CanSeeGroup(gs: GameState,
                 target: Group               = constraint.TEAM_GROUPS):
   """ Returns True if subject can see any of target
   """
-  return OR(*(CanSeeAgent(subject, agent) for agent in target.agents))
+  return POR(*(CanSeeAgent(subject, agent) for agent in target.agents))
```
Comment (daveey) [inline nmmo/task/base_predicates.py]: why POR not OR?

File: nmmo/task/predicate_api.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+      from nmmo.task.task_api import Task
+      task_cls = Task
+
+    if assignee is None:
+      # the new task is assigned to this predicate's subject
+      assignee = self._subject.agents
+
+    return task_cls(eval_fn=self, assignee=assignee, reward_multiplier=reward_multiplier)
+
+  def __and__(self, other):
```
Comment (daveey) [inline nmmo/task/predicate_api.py:145]: not a fan of the P prefix

File: nmmo/task/predicate_api.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+    return PAND(self, other)
+  def __or__(self, other):
+    return POR(self, other)
+  def __invert__(self):
+    return PNOT(self)
+  def __rshift__(self, other):
```
Comment (daveey) [inline nmmo/task/predicate_api.py]: IMPLY seems really weird for task predicates, do we really need it?

**Review by daveey — CHANGES_REQUESTED** [review]

File: nmmo/core/env.py (hunk 2 above)
Comment (kywch) [reply nmmo/core/env.py:91]: We need it, but not necessary in this format. Let me check back after seeing the syllabus integration.

File: nmmo/core/env.py (hunk 4 above)
Comment (kywch) [reply nmmo/core/env.py:312]: Previously, this was `gym_obs = {a: o.to_gym() for a,o in self.obs.items()}`, but it was changed to that to add in
```
      if self._task_encoding:
        gym_obs[a]['Task'] = self._encode_goal().get(a,np.zeros(self._task_embedding_size))
```

Currently, I'm not sure about the exact form of the task encoding. Hoping to get some input/specs as we start task conditioned learning very soon

File: nmmo/systems/skill.py (hunk 5 above)
Comment (kywch) [reply nmmo/systems/skill.py:272]: IMMORTAL is pretty much for performance testing. Perhaps change to PERFORMANCE_TEST?

File: nmmo/task/base_predicates.py (hunk 6 above)
Comment (kywch) [reply nmmo/task/base_predicates.py]: Removed!

File: nmmo/task/predicate_api.py (hunk 9 above)
Comment (kywch) [reply nmmo/task/predicate_api.py]: Removed.

File: nmmo/task/predicate_api.py (hunk 8 above)
Comment (kywch) [reply nmmo/task/predicate_api.py:145]: Removed.

File: nmmo/task/base_predicates.py (hunk 7 above)
Comment (kywch) [reply nmmo/task/base_predicates.py]: OR for Predicate. But I hear you, and am removing all P prefix.

File: nmmo/core/env.py (hunk 1 above)
Comment (kywch) [reply nmmo/core/env.py]: Can you please tell me more about it? Let me think about it again after sleep.

However, I also realized that the current `make_team_tasks(teams, task_spec)` doesn't need any info about the env. If one can pass in teams and task_spec, one can also use make_team_tasks to make task instances and pass them in. ... Then yes, the env doesn't need to know anything about making tasks or teams.

File: nmmo/core/env.py (hunk 3 above)
Comment (kywch) [reply nmmo/core/env.py]: nmmo_default_task has moved to the __init__, and this line was replaced with
```
      for task in self.tasks:
        task.reset()
```
not changing the existing tasks.

File: nmmo/core/env.py (hunk 1 above)
Comment (kywch) [reply nmmo/core/env.py]: Oh, I got it, and implemented as follows. But after writing it, I just thought... if one knows it all, why bother passing in these and not the instantiated tasks? Please let me know what you think.
```
    if new_tasks is not None:
      # providing an empty new_tasks [] is also possible
      self.tasks = new_tasks
    elif make_task_fn is not None:
      self.tasks = make_task_fn(**make_task_fn_kwargs)
    else:
      for task in self.tasks:
        task.reset()
```

File: nmmo/core/env.py [hunk 10]
```diff
@@ -41,15 +40,7 @@ def __init__(self,
...
-    # Default task: rewards 1 each turn agent is alive
-    self.tasks: List[Tuple[Task,float]] = None
-    self._task_encoding = None
-    self._task_embedding_size = -1
-    t = default_task(self.possible_agents)
-    self.change_task(t,
-                     embedding_size=self._task_embedding_size,
-                     task_encoding=self._task_encoding,
-                     reset=False)
+    self.tasks = nmmo_default_task(self.possible_agents)
```
Comment (daveey) [inline nmmo/core/env.py]: could you do task_api.make_nmmo_default_tasks()

File: nmmo/core/env.py [hunk 11]
```diff
@@ -131,38 +116,22 @@ def action_space(self, agent):
...
-    self._task_embedding_size = embedding_size
-    if reset:
-      self.reset(map_id=map_id, seed=seed, options=options)
-
   # TODO: This doesn't conform to the PettingZoo API
   # pylint: disable=arguments-renamed
-  def reset(self, map_id=None, seed=None, options=None):
+  def reset(self, map_id=None, seed=None, options=None,
+            new_tasks: List[Task]=None,
+            make_task_fn: Callable=None,
```
Comment (daveey) [inline nmmo/core/env.py]: we only need make_task_fn, the caller can define a lambda without having to pass kwargs. and they can always return [] if the want to clear the tasks. there's no need to accept new_tasks

File: nmmo/core/env.py [hunk 12]
```diff
@@ -442,27 +410,27 @@ def _compute_rewards(self, agents: List[AgentID], dones: Dict[AgentID, bool]):
...
+        task_rewards, task_infos = task.compute_rewards(self.game_state)
+        reward_cache[task] = (task_rewards, task_infos)
+      for agent_id, reward in task_rewards.items():
+        if agent_id in agents and agent_id not in dones:
+          rewards[agent_id] = rewards.get(agent_id,0) + reward
+          infos[agent_id]['task'][task.name] = task_infos[agent_id] # progress
+
+    # Remove rewards for dead agents
+    for agent_id in dones:
+      rewards[agent_id] = -1
```
Comment (daveey) [inline nmmo/core/env.py]: this seems wrong? dones only contains agents who died this turn. seems fine for them to get a reward. but agents that have already been dead should not be rewarded after death, right? 

or... maybe they should be, if their task is accomplished after they died

File: nmmo/core/realm.py [hunk 13]
```diff
@@ -104,6 +101,10 @@ def reset(self, map_id: int = None):
     Item.INSTANCE_ID = 0
     self.items = {}
 
+    if self._replay_helper is not None:
+      self._replay_helper.reset()
+      self._replay_helper.update() # capture the initial packet
```
Comment (daveey) [inline nmmo/core/realm.py]: should we just call update() in reset?

**Review by daveey — CHANGES_REQUESTED** [review]

File: nmmo/core/env.py (hunk 12 above)
Comment (kywch) [reply nmmo/core/env.py]: I see you are tinkering with this part too. I don't have any strong idea, so I'll just take it out.

File: nmmo/core/realm.py (hunk 13 above)
Comment (kywch) [reply nmmo/core/realm.py]: done.

File: nmmo/core/env.py (hunk 11 above)
Comment (kywch) [reply nmmo/core/env.py]: David, you suggest that we get rid of new_tasks, a list of task instances, and go only with make_task_fn, which can be
* make_task_fn = `lambda: return make_team_tasks(teams, task_spec)`   in case of the manual curriculum example
* or make_task_fn = 'lambda: return []' 

and inside reset(), `self.tasks = make_task_fn()` is called. Am I right? 

@jsuarez5341 how would the puffer like to handle this?

Comment (kywch) [reply nmmo/core/env.py]: I got the part that we don't need to pass in args for make_task_fn, so removed it.

And, I agree that new_tasks and make_task_fn are redundant. Well, ok ... it's pretty easy to put new_tasks in, so I'll go ahead only with make_task_fn.

File: nmmo/core/env.py (hunk 10 above)
Comment (kywch) [reply nmmo/core/env.py]: explicitly stated where the default task comes from.
